lib/auth/github.go

```diff
@@ -84,6 +91,124 @@ func (a *Server) upsertGithubConnector(ctx context.Context, connector types.Gith
+	version := modules.GetModules().BuildType()
+	if version == modules.BuildEnterprise {
+		return nil
+	}
+
+	orgs := make(map[string]struct{})
+	addOrg := func(org string) {
+		if len(userTeams) != 0 {
+			// Only check organizations that the user is a member of and
+			// that are specified in this auth connector
+			for _, team := range userTeams {
+				if org == team.Org.Login {
+					orgs[org] = struct{}{}
+				}
+			}
+		} else {
+			orgs[org] = struct{}{}
+		}
+	}
+
+	// Check each organization only once
+	for _, mapping := range conn.GetTeamsToLogins() {
+		addOrg(mapping.Organization)
+	}
+	for _, mapping := range conn.GetTeamsToRoles() {
+		addOrg(mapping.Organization)
+	}
+
+	if client == nil {
+		var err error
+		client, err = defaults.HTTPClient()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	for org := range orgs {
+		orgResult, err := orgCache.Get(ctx, org, func(ctx context.Context) (interface{}, error) {
+			return orgUsesExternalSSO(ctx, org, client)
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		usesSSO, ok := orgResult.(bool)
+		if !ok {
+			return trace.BadParameter("Expected bool from cache, got %T", orgResult)
+		}
+		if usesSSO {
+			return trace.AccessDenied(
+				"GitHub organization %s uses external SSO, please purchase a Teleport Enterprise license if you want to authenticate with this organization",
+				org,
+			)
+		}
+	}
+
+	return nil
+}
+
+// orgUsesExternalSSO returns true if the Github organization org
+// uses external SSO.
+func orgUsesExternalSSO(ctx context.Context, org string, client HTTPRequester) (bool, error) {
+	// A Github organization will have a "sso" page reachable if it
+	// supports external SSO. There doesn't seem to be any way to get this
+	// information from the Github REST API without being an owner of the
+	// Github organization, so check if this exists instead.
+	ssoURL := fmt.Sprintf("%s/%s/sso", githubOrgsURL, url.PathEscape(org))
+
+	const retries = 3
+	var resp *http.Response
+	for i := 0; i < retries; i++ {
+		ctx, cancel := context.WithTimeout(ctx, defaults.HTTPRequestTimeout)
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, ssoURL, nil)
+		cancel()
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+
+		var urlErr *url.Error
+		resp, err = client.Do(req)
+		if err == nil {
+			break
+		} else if errors.As(err, &urlErr) && urlErr.Timeout() {
+			if i == retries-1 {
+				// The connection timed out a couple of times in a row,
+				// stop trying and return the error.
+				return false, trace.ConnectionProblem(err, "Timed out trying to reach GitHub to check for organization external SSO.")
+			}
+			// Connection timed out, try to make the request again
+			continue
+		}
+		// Unknown error, don't try making any more requests
+		return false, trace.Wrap(err, "Unknown error trying to reach GitHub to check for organization external SSO")
+	}
+	err := resp.Body.Close()
```

Comment:
Shouldn't you be closing the response body inside the loop? If not, you'll be leaking resources in the case of a timeout, as only the last response will have its body closed.

Reply:
According to the docs: `On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned Response.Body is already closed.` That was my assumption I was operating off of, that the loop would exit only when there is no error so I don't need to close the response body otherwise.

Source: https://pkg.go.dev/net/http#Client.Do

Reply:
🤦‍♂️yep. makes sense.